pytensor/xtensor/shape.py

```diff
@@ -73,6 +75,135 @@ def stack(x, dim: dict[str, Sequence[str]] | None = None, **dims: Sequence[str])
+) -> tuple[str, ...]:
+    """Expand ellipsis in dimension permutation.
+
+    Parameters
+    ----------
+    dims : tuple[str, ...]
+        The dimension permutation, which may contain ellipsis
+    all_dims : tuple[str, ...]
+        All available dimensions
+    validate : bool, default True
+        Whether to check that all non-ellipsis elements in dims are valid dimension names.
+
+    Returns
+    -------
+    tuple[str, ...]
+        The expanded dimension permutation
+
+    Raises
+    ------
+    ValueError
+        If more than one ellipsis is present in dims.
+        If any non-ellipsis element in dims is not a valid dimension name and validate is True.
+    """
+    if dims == () or dims == (...,):
+        return tuple(reversed(all_dims))
+
+    if ... not in dims:
+        if validate:
+            invalid_dims = set(dims) - set(all_dims)
+            if invalid_dims:
+                raise ValueError(
+                    f"Invalid dimensions: {invalid_dims}. Available dimensions: {all_dims}"
+                )
+        return dims
+
+    if sum(d is ... for d in dims) > 1:
+        raise ValueError("an index can only have a single ellipsis ('...')")
+
+    pre = []
+    post = []
+    found = False
+    for d in dims:
+        if d is ...:
+            found = True
+        elif not found:
+            pre.append(d)
+        else:
+            post.append(d)
+    if validate:
+        invalid_dims = set(pre + post) - set(all_dims)
+        if invalid_dims:
+            raise ValueError(
+                f"Invalid dimensions: {invalid_dims}. Available dimensions: {all_dims}"
+            )
+    middle = [d for d in all_dims if d not in pre + post]
+    return tuple(pre + middle + post)
+
+
+class Transpose(XOp):
+    __props__ = ("dims", "missing_dims")
+
+    def __init__(
+        self,
+        dims: tuple[str, ...],
+        missing_dims: Literal["raise", "warn", "ignore"] = "raise",
+    ):
+        super().__init__()
+        self.dims = dims
+        self.missing_dims = missing_dims
+
+    def make_node(self, x):
+        x = as_xtensor(x)
+        dims = expand_ellipsis(
+            self.dims, x.type.dims, validate=(self.missing_dims == "raise")
+        )
+
+        # Handle missing dimensions based on missing_dims setting
+        if self.missing_dims in ("ignore", "warn"):
+            if self.missing_dims == "warn":
+                missing = set(dims) - set(x.type.dims)
+                if missing:
+                    warnings.warn(f"Dimensions {missing} do not exist in {x.type.dims}")
+            # Filter out dimensions that don't exist and add remaining ones
+            dims = tuple(d for d in dims if d in x.type.dims)
+            remaining_dims = tuple(d for d in x.type.dims if d not in dims)
+            dims = dims + remaining_dims
+        else:  # "raise"
+            if set(dims) != set(x.type.dims):
+                raise ValueError(f"Transpose dims {dims} must match {x.type.dims}")
```

Comment:
Isn't this repeating work already done in `expand_ellipsis`. If so, I still thing it's more reasonable to do it here than in a function called `expand_ellipsis`